### PR TITLE
SLLS-23 Code action to investigate taint vulnerability on the originating server

### PR DESCRIPTION
**Pull request by jblievremont**
This PR contains several changes:
- one new feature (code action to open a taint vulnerability in the browser)
- one missed feature (telemetry about taint vulnerabilities investigated locally)
- PM feedback from rounds of demos

### CODE REVIEW

File: src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java [hunk 1]
```diff
@@ -491,22 +506,29 @@ private static Range position(ServerIssueLocation issue) {
...
   static String message(ServerIssue issue) {
-    int nbFlows = issue.getFlows().size();
-    return String.format(MESSAGE_WITH_PLURALIZED_SUFFIX, issue.getMessage(), nbFlows, pluralize(nbFlows, "flow"));
+    if (issue.getFlows().isEmpty()) {
+      return issue.getMessage();
+    } else if (issue.getFlows().size() == 1) {
+      return buildMessageWithPluralizedSuffix(issue.getMessage(), issue.getFlows().get(0).locations().size(), ITEM_LOCATION);
+    } else {
+      return buildMessageWithPluralizedSuffix(issue.getMessage(), issue.getFlows().size(), ITEM_FLOW);
+    }
```
Comment (damien-urruty-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java:527]: Here we don't consider the case where we can several flows having each a single location (as we do for live issues), but maybe it's not relevant for taint vulnerabilities

File: src/main/java/org/sonarsource/sonarlint/ls/CommandManager.java [hunk 2]
```diff
@@ -183,6 +212,22 @@ private void handleOpenRuleDescriptionFromCodeActionCommand(ExecuteCommandParams
...
+  }
+
+  private void handleShowTaintVulnerabilityFlows(ExecuteCommandParams params) {
+    String issueKey = getAsString(params.getArguments().get(0));
+    String connectionId = getAsString(params.getArguments().get(1));
+    analysisManager.getTaintVulnerabilityByKey(issueKey)
+      .ifPresent(issue -> {
+        telemetry.taintVulnerabilitiesInvestigatedLocally();
+        client.showTaintVulnerability(ShowAllLocationsCommand.params(issue, connectionId, bindingManager::serverPathToFileUri));
+      });
```
Comment (damien-urruty-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/CommandManager.java:228]: what does it mean if the taint vulnerability is not present ? is this possible ? Should we give some feedback to the user ?

File: src/main/java/org/sonarsource/sonarlint/ls/Utils.java [hunk 3]
```diff
@@ -62,4 +65,21 @@ public static void interrupted(InterruptedException e) {
...
   }
+
+  public static String encodeUriComponent(String toEncode) {
+    try {
+      return URLEncoder.encode(toEncode, StandardCharsets.UTF_8.name());
+    } catch (UnsupportedEncodingException e) {
+      // Should not happen on a standard JVM
+      throw new IllegalStateException(e);
+    }
+  }
```
Comment (damien-urruty-sonarsource) [inline src/main/java/org/sonarsource/sonarlint/ls/Utils.java:76]: We already have StringUtils.urlEncode in SLCORE

**Review by damien-urruty-sonarsource — APPROVED** [review]
LGTM, just a few minor comments/questions

File: src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java (hunk 1 above)
Comment (jblievremont) [reply src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java:527]: Indeed, this case should only happen with local issues.

File: src/main/java/org/sonarsource/sonarlint/ls/Utils.java (hunk 3 above)
Comment (jblievremont) [reply src/main/java/org/sonarsource/sonarlint/ls/Utils.java:76]: Good catch :+1:

**Pull request merged**